modes/parkour/moderation.lua

```diff
@@ -0,0 +1,717 @@
+	end
+end
+
+onEvent("GameDataLoaded", function(data, fileid)
+	fileid = tonumber(fileid)
+	local action
+	local save
+	local len = #to_do
+
+	for index = 1, len do
+		action = to_do[index]
+		if files[action[1]] == fileid then
+			action[3](data)
+			save = action[2]
+			to_do[index] = false
+		end
+	end
+
+	for index = len, 1, -1 do
+		if not to_do[index] then
+			table.remove(to_do, index)
+		end
+	end
+
+	if data.lowmaps then
+		maps_loaded = true
+	end
+
+	if data.sanction then
+		sanctions_file = data.sanction
+
+		local now = os.time()
+		local count, to_remove = 0
+
+		-- wipe expired bans
+		for id, sanctiondata in pairs(data.sanction) do
+			if sanctiondata.time and sanctiondata.time > 2 and now >= sanctiondata.time then
+				if count == 0 then
+					count = 1
+					to_remove = {id}
+				else
+					count = count + 1
+					to_remove[count] = id
+				end
+			end
+		end
+
+		for index = 1, count do
+			local id = to_remove[index]
+			if data.sanction[id] then
+				data.sanction[id] = nil
+			end
+		end
+
+		local playerList = room.playerList
+		local id, banInfo, banDays
+
+		for player, pdata in next, players_file do
+			if playerList[player] and in_room[player] then
+				id = tostring(playerList[player].id)
+				banInfo = id and data.sanction[id]
+
+				if banInfo and banInfo.timestamp ~= pdata.lastsanction then
+					if banInfo.time == 1 then -- !ban
+						pdata.bancount = pdata.bancount + 1
+
+						if pdata.bancount == 1 then
+							pdata.banned = now + 86400000 -- 1 day
+						elseif pdata.bancount == 2 then
+							pdata.banned = now + 86400000 * 7
+						elseif pdata.bancount == 3 then
+							pdata.banned = now + 86400000 * 30
+						else
+							pdata.banned = 2 -- permanent ban
+						end
+					elseif banInfo.time == 0 then -- !punban
+						pdata.banned = nil
+					elseif banInfo.time == -1 then -- !unban
+						pdata.bancount = pdata.bancount - 1
+						pdata.banned = nil
+					else
+						pdata.banned = banInfo.time
+					end
+
+					pdata.lastsanction = banInfo.timestamp
+					pdata.bannedby = banInfo.info
+					
+					savePlayerData(player)
+					
+					if pdata.banned == nil then
+						data.sanction[id] = nil
+						save = true
+					else
+						banInfo.time = pdata.banned
+						save = true
+					end
+
+					local minutes = pdata.banned and math.floor((pdata.banned - os.time()) / 1000 / 60)
+					for moderator in pairs(room.playerList) do
+						if ranks.admin[moderator] or ranks.mod[moderator] then
+							sendBanLog(player, pdata.banned, moderator, minutes)
+						end
+					end
+				end
+
+				if pdata.banned and (pdata.banned == 2 or os.time() < pdata.banned) then
+					if pdata.banned == 2 then
+						translatedChatMessage("permbanned", player)
+					else
+						local minutes = math.floor((pdata.banned - os.time()) / 1000 / 60)
+						translatedChatMessage("tempbanned", player, minutes)
+					end
+				end
+			end
+		end
+	end
+
+	if save then
+		eventSavingFile(tostring(fileid), data)
```

Comment:
Do remember we're already triggering eventSavingFile in `files.lua`. I don't really see any situation where both would be triggered *in this revision*; but as a safe guard and somewhat future proofing, I would simply remove one of the two and add a single place where it is called.